src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -255,21 +303,112 @@ cdef class AioServer:
 
+    async def _server_main_loop(self,
+                                object server_started):
+        self._server.start()
+        cdef RPCState rpc_state
+        server_started.set_result(True)
+
+        while True:
+            # When shutdown begins, no more new connections.
+            if self._status != AIO_SERVER_STATUS_RUNNING:
+                break
+
+            rpc_state = await _server_call_request_call(
+                self._server,
+                self._cq,
+                self._loop)
+
+            self._loop.create_task(_handle_rpc(
+                self._generic_handlers,
+                rpc_state,
+                self._loop))
+
     async def start(self):
         if self._status == AIO_SERVER_STATUS_RUNNING:
             return
         elif self._status != AIO_SERVER_STATUS_READY:
             raise RuntimeError('Server not in ready state')
 
         self._status = AIO_SERVER_STATUS_RUNNING
-        loop = asyncio.get_event_loop()
-        loop.create_task(_server_start(
-            self._server,
-            self._cq,
-            self._generic_handlers,
-        ))
+        cdef object server_started = self._loop.create_future()
+        self._serving_task = self._loop.create_task(self._server_main_loop(server_started))
+        # Needs to explicitly wait for the server to start up.
+        # Otherwise, the actual start time of the server is un-controllable.
+        await server_started
+
+    async def _start_shutting_down(self):
+        """Prepares the server to shutting down.
+
+        This coroutine function is NOT coroutine-safe.
+        """
+        # The shutdown callback won't be called until there is no live RPC.
+        grpc_server_shutdown_and_notify(
+            self._server.c_server,
+            self._cq._cq,
+            self._shutdown_callback_wrapper.c_functor())
+
+        # Ensures the serving task (coroutine) exits.
+        try:
+            await self._serving_task
+        except _RequestCallError:
+            pass
+
+    async def shutdown(self, grace):
+        """Gracefully shutdown the C-Core server.
+
+        Application should only call shutdown once.
+
+        Args:
+          grace: An optional float indicating the length of grace period in
+            seconds.
+        """
+        if self._status == AIO_SERVER_STATUS_READY or self._status == AIO_SERVER_STATUS_STOPPED:
+            return
+
+        async with self._shutdown_lock:
+            if self._status == AIO_SERVER_STATUS_RUNNING:
+                await self._start_shutting_down()
+                self._server.is_shutting_down = True
+                self._status = AIO_SERVER_STATUS_STOPPING
+
+        if grace is None:
+            # Directly cancels all calls
+            grpc_server_cancel_all_calls(self._server.c_server)
+            await self._shutdown_completed
+        else:
+            try:
+                await asyncio.wait_for(asyncio.shield(self._shutdown_completed), grace)
```

Comment:
`_shutdown_completed` is a future that completes once the shutdown-complete-event propagates from Core. The destroy of Core server object will only be safe if that event is observed in wrapper layer. So, we still want the future to end properly with a `Done` state not `Cancelled` state.

Reply:
Umm, but what `shield` does is protect the coroutine to not be canceled because of the parent one was, I don't get the point why it is useful in this situation. 

Consider a situation where `shield` could be useful:

```python
async def handler(self, request):
    if requets.host not in cache:
        ip = await asyncio.wait_for( asyuncio.shield(dns_resolution_and_cache(request.hostname)))
```
In the previous scenario no matter what we want to finish the resolution and store the result in a cache.

How is this pattern related to the `shutdown` coroutine called by the user and the shielding of the `_shutdown_completed`? 

I'm a bit lost, sorry.

Reply:
There are two questions need to be answered:

### What does the `asyncio.shield` provides?
In `asyncio` world, if parent coroutine is cancelled, child coroutine, awaited coroutines will also be cancelled. Just as you said. But there is a slightly different case for `asyncio.wait_for` here, it would cancel all supplied coroutines if timeout occurs. So, adding the `asyncio.shield` here protects supplied future being cancelled by `asyncio.wait_for`. A glance at the [doc](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for):

> Wait for the aw awaitable to complete with a timeout.
> 
> If aw is a coroutine it is automatically scheduled as a Task.
> 
> timeout can either be None or a float or int number of seconds to wait for. If timeout is None, block until the future completes.
> 
> If a timeout occurs, it cancels the task and raises asyncio.TimeoutError.
> 
> To avoid the task cancellation, wrap it in shield().

### Why do we need it?
`_shutdown_completed` future is bound to the execution of server shutdown callback. This callback indicates the readiness of Core to destroy C server object safely. I'm taking the most conservative route here, that we must make sure the Core is ready, before we proceed to the server destruction (otherwise, undefined behavior, segfault, abort, might happen). So, we cannot cancel the `_shutdown_completed` future, and want to ensure its preciseness.

Reply:
Got it, thanks for the clear explanation!